lib/DWARFImporter/DWARFImporter.cpp

```diff
@@ -0,0 +1,189 @@
+
+using namespace swift;
+
+std::unique_ptr<DWARFImporter>
+DWARFImporter::create(ASTContext &ctx,
+                      const ClangImporterOptions &importerOpts,
+                      DependencyTracker *tracker) {
+  std::unique_ptr<DWARFImporter> importer{
+    new DWARFImporter(ctx, importerOpts, tracker)
+  };
+  return importer;
+}
+
+class DWARFModuleUnit final : public LoadedFile {
+
+  ~DWARFModuleUnit() = default;
+
+public:
+  DWARFModuleUnit(ModuleDecl &M)
+      : LoadedFile(FileUnitKind::DWARFModule, M) {}
+
+  virtual bool isSystemModule() const override { return false; }
+
+  virtual void
+  lookupValue(ModuleDecl::AccessPathTy accessPath, DeclName name,
+              NLKind lookupKind,
+              SmallVectorImpl<ValueDecl *> &results) const override {}
+
+  virtual TypeDecl *
+  lookupNestedType(Identifier name,
+                   const NominalTypeDecl *baseType) const override {
+    return nullptr;
+  }
+
+  virtual void lookupVisibleDecls(ModuleDecl::AccessPathTy accessPath,
+                                  VisibleDeclConsumer &consumer,
+                                  NLKind lookupKind) const override {}
+
+  virtual void
+  lookupClassMembers(ModuleDecl::AccessPathTy accessPath,
+                     VisibleDeclConsumer &consumer) const override {}
+
+  virtual void
+  lookupClassMember(ModuleDecl::AccessPathTy accessPath, DeclName name,
+                    SmallVectorImpl<ValueDecl *> &decls) const override {}
+
+  void lookupObjCMethods(
+      ObjCSelector selector,
+      SmallVectorImpl<AbstractFunctionDecl *> &results) const override {}
+
+  virtual void
+  getTopLevelDecls(SmallVectorImpl<Decl *> &results) const override {}
+
+  virtual void
+  getDisplayDecls(SmallVectorImpl<Decl *> &results) const override {}
+
+  virtual void
+  getImportedModules(SmallVectorImpl<ModuleDecl::ImportedModule> &imports,
+                     ModuleDecl::ImportFilter filter) const override {}
+
+  virtual void getImportedModulesForLookup(
+      SmallVectorImpl<ModuleDecl::ImportedModule> &imports) const override {};
+
+  virtual void collectLinkLibraries(
+      ModuleDecl::LinkLibraryCallback callback) const override {};
+
+  Identifier
+  getDiscriminatorForPrivateValue(const ValueDecl *D) const override {
+    llvm_unreachable("no private decls in Clang modules");
+  }
+
+  virtual StringRef getFilename() const override { return ""; }
+
+  virtual const clang::Module *getUnderlyingClangModule() const override {
+    return nullptr;
+  }
+
+  static bool classof(const FileUnit *file) {
+    return file->getKind() == FileUnitKind::DWARFModule;
+  }
+  static bool classof(const DeclContext *DC) {
+    return isa<FileUnit>(DC) && classof(cast<FileUnit>(DC));
+  }
+};
+
+class DWARFImporter::Implementation {
+public:
+  ASTContext &SwiftContext;
+  clang::ASTContext *ClangASTContext = nullptr;
+  clang::CompilerInstance *Instance = nullptr;
+  clang::Preprocessor *PP = nullptr;
+  clang::Sema *Sema = nullptr;
```

Comment:
I'm worried about this part. Are there going to be two separate clang::CompilerInstances in the same Swift ASTContext?

Reply:
I haven't thought about this in depth, but most likely LLDB (which is creating the _real_ ClangImporter, too) should probably pass in the real ClangImporter's CompilerInstance etc. I was originally worried that I wouldn't want to mix the two, but LLDB already knows how to tear down and recreate a broken SwiftASTContext, so that approach would probably work.

Reply:
When we originally talked about this, I figured you'd have DWARFModuleLoader subclass ClangModuleLoader, so that you'd really only have one, and the DWARF parts would just go first. Your current approach might be better (and does have a nice "separation of concerns" aspect), but if you hadn't thought of that one it might be worth considering!

Reply:
I will revisit this once the actual implementation is more fleshed out. For the bringup the current approach makes it cleaner to make the DWARF part optional.